README.md
# docker-postgresql

[![Build Status](https://travis-ci.com/madharjan/docker-postgresql.svg?branch=master)](https://travis-ci.com/madharjan/docker-postgresql)
[![Layers](https://images.microbadger.com/badges/image/madharjan/docker-postgresql.svg)](http://microbadger.com/images/madharjan/docker-postgresql)

Docker container for PostgreSQL Server based on [madharjan/docker-base](https://github.com/madharjan/docker-base/)

## Features

* Environment variables to create database, user and set password
* Bats [bats-core/bats-core](https://github.com/bats-core/bats-core) based test cases

## PostgreSQL Server 9.5 (docker-postgresql)

### Environment

| Variable             | Default      | Example        |
|----------------------|--------------|----------------|
| DISABLE_POSTGRESQL   | 0            | 1 (to disable) |
| POSTGRESQL_DATABASE  | postgres     | mydb           |
| POSTGRESQL_USERNAME  | postgres     | myuser         |
| POSTGRESQL_PASSWORD  |              | mypass         |

## Build

### Clone this project

```bash
git clone https://github.com/madharjan/docker-postgresql
cd docker-postgresql
```

### Build Containers

```bash
# login to DockerHub
docker login

# build
make

# tests
make run
make tests
make clean

# tag
make tag_latest

# release
make release
```

### Tag and Commit to Git

```bash
git tag 9.5
git push origin 9.5
```

## Run Container

### Prepare folder on host for container volumes

```bash
sudo mkdir -p /opt/docker/postgresql/etc/
sudo mkdir -p /opt/docker/postgresql/lib/
sudo mkdir -p /opt/docker/postgresql/log/
```

### Run `docker-postgresql`

```bash
docker stop postgresql
docker rm postgresql

docker run -d \
  -e POSTGRESQL_DATABASE=mydb \
  -e POSTGRESQL_USERNAME=myuser \
  -e POSTGRESQL_PASSWORD=mypass \
  -p 5432:5432 \
  -v /opt/docker/postgresql/etc:/etc/postgresql/9.5/main \
  -v /opt/docker/postgresql/lib:/var/lib/postgresql/9.5/main \
  -v /opt/docker/postgresql/log:/var/log/postgresql \
  --name postgresql \
  madharjan/docker-postgresql:9.5
```

## Run via Systemd

### Systemd Unit File - basic example

```txt
[Unit]
Description=PostgreSQL Server

After=docker.service

[Service]
TimeoutStartSec=0

ExecStartPre=-/bin/mkdir -p /opt/docker/postgresql/etc
ExecStartPre=-/bin/mkdir -p /opt/docker/postgresql/lib
ExecStartPre=-/bin/mkdir -p /opt/docker/postgresql/log
ExecStartPre=-/usr/bin/docker stop postgresql
ExecStartPre=-/usr/bin/docker rm postgresql
ExecStartPre=-/usr/bin/docker pull madharjan/docker-postgresql:9.5

ExecStart=/usr/bin/docker run \
  -e POSTGRESQL_DATABASE=mydb \
  -e POSTGRESQL_USERNAME=myuser \
  -e POSTGRESQL_PASSWORD=mypass \
  -p 5432:5432 \
  -v /opt/docker/postgresql/etc:/etc/postgresql/etc/9.5/main \
  -v /opt/docker/postgresql/lib:/var/lib/postgresql/9.5/main \
  -v /opt/docker/postgresql/log:/var/log/postgresql \
  --name postgresql \
  madharjan/docker-postgresql:9.5

ExecStop=/usr/bin/docker stop -t 2 postgresql

[Install]
WantedBy=multi-user.target
```

### Generate Systemd Unit File

| Variable                 | Default          | Example                                                          |
|--------------------------|------------------|------------------------------------------------------------------|
| PORT                     |                  | 5432                                                             |
| VOLUME_HOME              | /opt/docker      | /opt/data                                                        |
| VERSION                  | 9.5              | latest                                                           |
| NAME                     | postgresql       | docker-postgresql                                                |
| POSTGRESQL_DATABASE      | postgres         | mydb                                                             |
| POSTGRESQL_USERNAME      | postgres         | user                                                             |
| POSTGRESQL_PASSWORD      |                  | pass                                                             |

```bash
docker run --rm \
  -e PORT=5432 \
  -e VOLUME_HOME=/opt/docker \
  -e VERSION=9.5 \
  -e NAME=docker-postgresql \
  -e POSTGRESQL_DATABASE=mydb \
  -e POSTGRESQL_USERNAME=user \
  -e POSTGRESQL_PASSWORD=pass \
  madharjan/docker-postgresql:9.5 \
  postgresql-systemd-unit | \
  sudo tee /etc/systemd/system/postgresql.service

sudo systemctl enable postgresql
sudo systemctl start postgresql
```


Changelog.md
## 9.5 

### Changes

* Based on madharjan/docker-base:16.04
* Using Postgres SQL 9.5.16

## 9.3 

### Changes

* Based on madharjan/docker-base:14.04
* Using Postgres SQL 9.3

services/20-postgresql.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

pgsql_client()
{
  local cmd="$1"
  su - ${POSTGRESQL_USER} -c "/usr/bin/psql -c \"${cmd}\"" 2>&1
}

DISABLE_POSTGRESQL=${DISABLE_POSTGRESQL:-0}

if [ ! "${DISABLE_POSTGRESQL}" -eq 0 ]; then
  touch /etc/service/postgresql/down
else
  rm -f /etc/service/postgresql/down
fi

POSTGRESQL_VERSION=${POSTGRESQL_VERSION:-9.5}

POSTGRESQL_DATABASE=${POSTGRESQL_DATABASE:-postgres}
POSTGRESQL_USERNAME=${POSTGRESQL_USERNAME:-postgres}
POSTGRESQL_PASSWORD=${POSTGRESQL_PASSWORD:-}
POSTGRESQL_ENCODING=${POSTGRESQL_ENCODING:-UNICODE}

POSTGRESQL_USER=postgres
POSTGRESQL_GROUP=postgres

POSTGRESQL_CONF_DIR=${POSTGRESQL_CONF_DIR:-/etc/postgresql/${POSTGRESQL_VERSION}/main}
POSTGRESQL_DATA_DIR=${POSTGRESQL_DATA_DIR:-/var/lib/postgresql/${POSTGRESQL_VERSION}/main}
POSTGRESQL_BIN_DIR=${POSTGRESQL_BIN_DIR:-/usr/lib/postgresql/${POSTGRESQL_VERSION}/bin}

mkdir -p ${POSTGRESQL_DATA_DIR}
chmod -R 0700 ${POSTGRESQL_DATA_DIR}
chown -R ${POSTGRESQL_USER}:${POSTGRESQL_GROUP} ${POSTGRESQL_DATA_DIR}

cd ${POSTGRESQL_DATA_DIR}

if [ ! -s "$POSTGRESQL_DATA_DIR/PG_VERSION" ]; then
  echo "Initializing database ..."
  pg_createcluster -u ${POSTGRESQL_USER} -g ${POSTGRESQL_GROUP} -d ${POSTGRESQL_DATA_DIR} ${POSTGRESQL_VERSION} main

  if [ "$POSTGRESQL_PASSWORD" ]; then
    { echo; echo "host all all 0.0.0.0/0 md5"; } >> "${POSTGRESQL_CONF_DIR}/pg_hba.conf"
    { echo; echo "listen_addresses='*'"; } >> "${POSTGRESQL_CONF_DIR}/postgresql.conf"
  else
    cat >&2 <<-'EOWARN'
****************************************************
WARNING: No password has been set for the database.
         Use "-e POSTGRESQL_PASSWORD=password" to
         set it in "docker run".
****************************************************
EOWARN
  fi

  pg_ctlcluster ${POSTGRESQL_VERSION} main start

  if [ "$POSTGRESQL_DATABASE" != 'postgres' ]; then
    echo "Creating database ..."
    pgsql_client "CREATE DATABASE ${POSTGRESQL_DATABASE} ENCODING '${POSTGRESQL_ENCODING}';"
  fi

  echo "Granting access to database \"${POSTGRESQL_DATABASE}\" for user \"${POSTGRESQL_USERNAME}\"..."
  if [ "$POSTGRESQL_USERNAME" = 'postgres' ]; then
      OPERATION='ALTER'
  else
      OPERATION='CREATE'
  fi
  pgsql_client "${OPERATION} USER ${POSTGRESQL_USERNAME} WITH SUPERUSER PASSWORD '$POSTGRESQL_PASSWORD';"
  pgsql_client "GRANT ALL PRIVILEGES ON DATABASE ${POSTGRESQL_DATABASE} TO ${POSTGRESQL_USERNAME};"

  pg_ctlcluster ${POSTGRESQL_VERSION} main stop

  echo 'PostgreSQL init process complete; ready for start up.'
fi


services/postgresql-startup.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

pgsql_client()
{
  local cmd="$1"
  sudo su - ${POSTGRESQL_USER} -c "/usr/bin/psql -c \"${cmd}\"" 2>&1
}

DISABLE_POSTGRESQL=${DISABLE_POSTGRESQL:-0}

if [ ! "${DISABLE_POSTGRESQL}" -eq 0 ]; then
  touch /etc/service/postgresql/down
else
  rm -f /etc/service/postgresql/down
fi

POSTGRESQL_VERSION=${POSTGRESQL_VERSION:-9.3}

POSTGRESQL_DATABASE=${POSTGRESQL_DATABASE:-postgres}
POSTGRESQL_USERNAME=${POSTGRESQL_USERNAME:-postgres}
POSTGRESQL_PASSWORD=${POSTGRESQL_PASSWORD:-}
POSTGRESQL_ENCODING=${POSTGRESQL_ENCODING:-UNICODE}

POSTGRESQL_USER=postgres
POSTGRESQL_GROUP=postgres

POSTGRESQL_CONF_DIR=${POSTGRESQL_CONF_DIR:-/etc/postgresql/${POSTGRESQL_VERSION}/main}
POSTGRESQL_DATA_DIR=${POSTGRESQL_DATA_DIR:-/var/lib/postgresql/${POSTGRESQL_VERSION}/main}
POSTGRESQL_BIN_DIR=${POSTGRESQL_BIN_DIR:-/usr/lib/postgresql/${POSTGRESQL_VERSION}/bin}

mkdir -p ${POSTGRESQL_DATA_DIR}
chmod -R 0700 ${POSTGRESQL_DATA_DIR}
chown -R ${POSTGRESQL_USER}:${POSTGRESQL_GROUP} ${POSTGRESQL_DATA_DIR}

cd ${POSTGRESQL_DATA_DIR}

if [ ! -s "$POSTGRESQL_DATA_DIR/PG_VERSION" ]; then
  echo "Initializing database ..."
  pg_createcluster -u ${POSTGRESQL_USER} -g ${POSTGRESQL_GROUP} -d ${POSTGRESQL_DATA_DIR} ${POSTGRESQL_VERSION} main

  if [ "$POSTGRESQL_PASSWORD" ]; then
    { echo; echo "host all all 0.0.0.0/0 md5"; } >> "${POSTGRESQL_CONF_DIR}/pg_hba.conf"
    { echo; echo "listen_addresses='*'"; } >> "${POSTGRESQL_CONF_DIR}/postgresql.conf"
  else
    cat >&2 <<-'EOWARN'
****************************************************
WARNING: No password has been set for the database.
         Use "-e POSTGRESQL_PASSWORD=password" to
         set it in "docker run".
****************************************************
EOWARN
  fi

  pg_ctlcluster ${POSTGRESQL_VERSION} main start

  if [ "$POSTGRESQL_DATABASE" != 'postgres' ]; then
    echo "Creating database ..."
    pgsql_client "CREATE DATABASE ${POSTGRESQL_DATABASE} ENCODING '${POSTGRESQL_ENCODING}';"
  fi

  echo "Granting access to database \"${POSTGRESQL_DATABASE}\" for user \"${POSTGRESQL_USERNAME}\"..."
  if [ "$POSTGRESQL_USERNAME" = 'postgres' ]; then
      OPERATION='ALTER'
  else
      OPERATION='CREATE'
  fi
  pgsql_client "${OPERATION} USER ${POSTGRESQL_USERNAME} WITH SUPERUSER PASSWORD '$POSTGRESQL_PASSWORD';"
  pgsql_client "GRANT ALL PRIVILEGES ON DATABASE ${POSTGRESQL_DATABASE} TO ${POSTGRESQL_USERNAME};"

  pg_ctlcluster ${POSTGRESQL_VERSION} main stop

  echo 'PostgreSQL init process complete; ready for start up.'
fi
